File: 프로그래머스/3/43163. 단어 변환/단어 변환.js
function solution(begin, target, words) {
    if (!words.includes(target)) return 0;

    function isConnected(word1, word2) {
        let diff = 0;
        for (let i = 0; i < word1.length; i++) {
            if (word1[i] !== word2[i]) diff++;
            if (diff > 1) return false;
        }
        return diff === 1;
    }

    let min = Infinity;
    
    function dfs(current, count, visited) {
        if (current === target) {
            min = Math.min(min, count);
            return;
        }
        for (let i = 0; i < words.length; i++) {
            if (!visited[i] && isConnected(current, words[i])) {
                visited[i] = true;
                dfs(words[i], count + 1, visited);
                visited[i] = false;
            }
        }
    }

    dfs(begin, 0, Array(words.length).fill(false));
    return min === Infinity ? 0 : min;
}

File: 프로그래머스/3/43163. 단어 변환/단어 변환.py
def solution(begin, target, words):
    
    if target not in words:
        return 0

    N = len(words)
    visited = [False] * N
    
    def diff(word1, word2):
        count = 0
        for i in range(N):
            if word1[i: i+1] != word2[i: i+1]:
                count += 1
        return count == 1
    
    count = [0]
    
    def dfs(word, cnt):
        if (word == target):
            count[0] = cnt
            return
        
        for i in range(N):
            if not visited[i] and diff(word, words[i]):
                visited[i] = True
                dfs(words[i], cnt + 1)
                visited[i] = False
    
    dfs(begin, 0)
    
    return count[0]